File: cpp/20.valid-parentheses.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 *
 * https://leetcode-cn.com/problems/valid-parentheses/description/
 *
 * algorithms
 * Easy (43.34%)
 * Total Accepted:    484.1K
 * Total Submissions: 1.1M
 * Testcase Example:  '"()"'
 *
 * 给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串，判断字符串是否有效。
 * 
 * 有效字符串需满足：
 * 
 * 
 * 左括号必须用相同类型的右括号闭合。
 * 左括号必须以正确的顺序闭合。
 * 
 * 
 * 注意空字符串可被认为是有效字符串。
 * 
 * 示例 1:
 * 
 * 输入: "()"
 * 输出: true
 * 
 * 
 * 示例 2:
 * 
 * 输入: "()[]{}"
 * 输出: true
 * 
 * 
 * 示例 3:
 * 
 * 输入: "(]"
 * 输出: false
 * 
 * 
 * 示例 4:
 * 
 * 输入: "([)]"
 * 输出: false
 * 
 * 
 * 示例 5:
 * 
 * 输入: "{[]}"
 * 输出: true
 * 
 */
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> couple = {
            {'(', ')'}, {'[', ']'}, {'{', '}'},
        };

        stack<char> st;
        for (int i = 0; i < s.size(); ++i) {
            if (couple.find(s[i]) != couple.end()) {
                st.push(s[i]);
            } else if (!st.empty() && couple[st.top()] == s[i]) {
                st.pop();
            } else {
                return false;
            }
        }

        return st.empty();
    }
};

File: golang/20.valid-parentheses.go
package main
/*
 * [20] Valid Parentheses
 *
 * https://leetcode.com/problems/valid-parentheses/description/
 *
 * algorithms
 * Easy (34.22%)
 * Total Accepted:    354.5K
 * Total Submissions: 1M
 * Testcase Example:  '"()"'
 *
 * Given a string containing just the characters '(', ')', '{', '}', '[' and
 * ']', determine if the input string is valid.
 * 
 * An input string is valid if:
 * 
 * 
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 * 
 * 
 * Note that an empty string is also considered valid.
 * 
 * Example 1:
 * 
 * 
 * Input: "()"
 * Output: true
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: "()[]{}"
 * Output: true
 * 
 * 
 * Example 3:
 * 
 * 
 * Input: "(]"
 * Output: false
 * 
 * 
 * Example 4:
 * 
 * 
 * Input: "([)]"
 * Output: false
 * 
 * 
 * Example 5:
 * 
 * 
 * Input: "{[]}"
 * Output: true
 * 
 * 
 */

func isValid(s string) bool {
	stack := []rune {}
	pair := map[rune]rune { ']': '[', ')': '(', '}':'{' }

	for _, ch := range s {
		switch ch {
		case '[', '{', '(':
			stack = append(stack, ch)
		default:
			N := len(stack)
			if N == 0 {
				return false
			}

			last := stack[N-1]
			stack = stack[:N-1]

			if last != pair[ch] {
				return false
			}
		}
	}

	return len(stack) == 0
}
